Applications/Spire/Include/Spire/Ui/AssociativeValueModel.hpp

```diff
@@ -0,0 +1,242 @@
+      /**
+       * Returns the AssociativeValueModel's current value.
+       * @return The current value.
+       */
+      const Type& get_current() const override;
+
+      /**
+       * Sets the current value, iff a model is associated with the value.
+       * @param value The current value.
+       * @return Acceptable if the value was set successfully, Invalid
+       *         otherwise.
+       */
+      QValidator::State set_current(const Type& value) override;
+
+      /** Connects a slot to the current signal. */
+      boost::signals2::connection connect_current_signal(
+        const typename CurrentSignal::slot_type& slot) const override;
+
+    private:
+      mutable CurrentSignal m_current_signal;
+      Type m_current;
+      std::unique_ptr<Type> m_default_value;
+      std::unordered_map<Type, Details::AssociatedBooleanModel, Hash> m_models;
+      bool m_is_blocked;
+
+      void set_associated_model_value(const Type& value, bool model_value);
+      void on_current(const Type& value, bool is_selected);
+  };
+
+  template<typename T, typename U>
+  AssociativeValueModel<T, U>::AssociativeValueModel()
+    : m_is_blocked(false) {}
+
+  template<typename T, typename U>
+  void AssociativeValueModel<T, U>::associate(
+      const std::shared_ptr<BooleanModel>& model, const T& value) {
+    if(m_models.find(value) != m_models.end()) {
+      return;
+    }
+    if(model->get_current()) {
+      model->set_current(false);
+    }
+    m_models.insert_or_assign(value, Details::AssociatedBooleanModel{model,
+      model->connect_current_signal([=] (auto is_selected) {
+        on_current(value, is_selected);
+      })});
+    if(get_state() == QValidator::Invalid) {
+      set_current(value);
+    }
+  }
+
+  template<typename T, typename U>
+  std::shared_ptr<BooleanModel> AssociativeValueModel<T, U>::make_association(
+      const T& value) {
+    auto model = std::make_shared<LocalBooleanModel>(false);
+    associate(model, value);
+    return model;
+  }
+
+  template<typename T, typename U>
+  std::shared_ptr<BooleanModel> AssociativeValueModel<T, U>::find(
+      const T& value) const {
+    auto iterator = m_models.find(value);
+    if(iterator == m_models.end()) {
+      return nullptr;
+    }
+    return iterator->second.m_model;
+  }
+
+  template<typename T, typename U>
+  QValidator::State AssociativeValueModel<T, U>::set_default_value(
+      const Type& value) {
+    m_default_value.reset();
+    if(m_models.find(value) != m_models.end()) {
+      m_default_value = std::make_unique<Type>(value);
+      return QValidator::Acceptable;
+    }
+    return QValidator::Invalid;
+  }
+
+  template<typename T, typename U>
+  QValidator::State AssociativeValueModel<T, U>::get_state() const {
+    if(m_models.find(m_current) != m_models.end()) {
+      return QValidator::Acceptable;
+    }
+    return QValidator::Invalid;
+  }
+
+  template<typename T, typename U>
+  const T& AssociativeValueModel<T, U>::get_current() const {
```

Comment:
```const T&``` -> ```const typename AssociativeValueModel<T, U>::Type&```

Reply:
What's the advantage of this?

Reply:
It matches the signature of the method exactly. The method signature doesn't use ```T```, it uses ```Type```.